File: clients/js/apex_kv.js
/**
 * apex-kv — Node.js client for APEX-KV distributed key-value store
 *
 * Works in Node.js 18+ (uses built-in `net` module).
 *
 * @example
 * const { ApexClient } = require('./apex_kv');
 * const client = new ApexClient('127.0.0.1', 7001);
 * await client.put('hello', 'world');
 * const val = await client.get('hello');   // 'world'
 * await client.del('hello');
 * client.close();
 */
'use strict';

const net = require('net');

// Wire protocol
const MAGIC       = 0xA9ECBD10;
const HDR_SIZE    = 16;
const Cmd = {
  GET:          0x01, PUT:    0x02, DEL:  0x03, PING: 0x04, METRICS: 0x05,
  OK:           0x10, VALUE:  0x11, NOT_FOUND: 0x12, ERROR: 0x13,
  REDIRECT:     0x14, PONG:   0x15, METRICS_RESP: 0x16,
};

class ApexError      extends Error {}
class ApexNotFound   extends ApexError { constructor() { super('Key not found'); } }
class ApexNotLeader  extends ApexError { constructor(addr) { super(`Not leader, redirect to ${addr}`); this.leaderAddr = addr; } }

class ApexClient {
  /**
   * @param {string} host
   * @param {number} port
   * @param {{ autoRedirect?: boolean, connectTimeout?: number }} [opts]
   */
  constructor(host, port, opts = {}) {
    this._host          = host;
    this._port          = port;
    this._autoRedirect  = opts.autoRedirect  ?? true;
    this._connectTimeout = opts.connectTimeout ?? 5000;
    this._seq           = 0;
    this._socket        = null;
    this._recvBuf       = Buffer.alloc(0);
    this._pendingReqs   = [];
    this._connected     = false;
    this._connecting    = null;
  }

  /** Connect (called automatically on first use). */
  async connect() {
    if (this._connected) return;
    if (this._connecting) return this._connecting;
    this._connecting = new Promise((resolve, reject) => {
      const s = net.createConnection({ host: this._host, port: this._port });
      s.setNoDelay(true);
      const timer = setTimeout(() => { s.destroy(); reject(new ApexError('Connect timeout')); }, this._connectTimeout);
      s.once('connect', () => { clearTimeout(timer); this._socket = s; this._connected = true; resolve(); });
      s.once('error', err => { clearTimeout(timer); reject(new ApexError(`Connect: ${err.message}`)); });
      s.on('data', chunk => this._onData(chunk));
      s.on('close', () => { this._connected = false; this._socket = null; this._rejectAll(new ApexError('Connection closed')); });
    });
    return this._connecting;
  }

  /** Close the connection. */
  close() {
    if (this._socket) { this._socket.destroy(); this._socket = null; this._connected = false; }
  }

  /** Ping the node. Returns true if alive. */
  async ping() {
    await this.connect();
    const [cmd] = await this._rpc(Cmd.PING);
    return cmd === Cmd.PONG;
  }

  /**
   * Get a value by key.
   * @returns {Promise<string|null>} value or null if not found
   */
  async get(key) {
    await this.connect();
    const payload = this._encStr(key);
    const [cmd, resp] = await this._rpc(Cmd.GET, payload);
    if (cmd === Cmd.VALUE)     return this._decStr(resp);
    if (cmd === Cmd.NOT_FOUND) return null;
    if (cmd === Cmd.ERROR)     throw new ApexError(this._decStr(resp));
    return null;
  }

  /**
   * Put a key-value pair.
   * Automatically follows leader redirects.
   */
  async put(key, value) {
    await this.connect();
    const payload = Buffer.concat([this._encStr(key), this._encStr(value)]);
    const [cmd, resp] = await this._rpc(Cmd.PUT, payload, /* followRedirect */ true);
    if (cmd !== Cmd.OK) throw new ApexError(`PUT failed (${cmd}): ${this._decStr(resp)}`);
  }

  /**
   * Delete a key.
   * @returns {Promise<boolean>} true if key existed
   */
  async del(key) {
    await this.connect();
    const payload = this._encStr(key);
    const [cmd] = await this._rpc(Cmd.DEL, payload, true);
    return cmd === Cmd.OK;
  }

  /** Get value or return a default. */
  async getOrDefault(key, def = '') {
    const v = await this.get(key);
    return v !== null ? v : def;
  }

  /** Fetch server metrics. */
  async metrics() {
    await this.connect();
    const [cmd, resp] = await this._rpc(Cmd.METRICS);
    if (cmd !== Cmd.METRICS_RESP || resp.length < 32) return {};
    return {
      opsGet:     resp.readBigUInt64BE(0),
      opsPut:     resp.readBigUInt64BE(8),
      p99GetUs:   resp.readBigUInt64BE(16),
      p99PutUs:   resp.readBigUInt64BE(24),
    };
  }

  /**
   * Simple polling watch — calls onChange when any of `keys` change.
   * @param {string[]} keys
   * @param {function} onChange
   * @param {{ intervalMs?: number }} [opts]
   * @returns {{ stop: function }}
   */
  watch(keys, onChange, opts = {}) {
    const interval = opts.intervalMs ?? 500;
    const snapshot = {};
    let running = true;
    const poll = async () => {
      for (const key of keys) {
        try {
          const val = await this.get(key);
          const prev = snapshot[key];
          if (val !== prev) {
            snapshot[key] = val;
            onChange({ key, value: val, deleted: val === null, prev });
          }
        } catch { /* ignore transient errors */ }
      }
      if (running) setTimeout(poll, interval);
    };
    poll();
    return { stop() { running = false; } };
  }

  // ── Internal ────────────────────────────────────────────────────────────────

  async _rpc(cmd, payload = Buffer.alloc(0), followRedirect = false) {
    const seq = ++this._seq;
    const hdr = Buffer.alloc(HDR_SIZE);
    hdr.writeUInt32BE(MAGIC, 0);
    hdr.writeUInt8(cmd, 4);
    hdr.writeUInt8(0, 5);
    hdr.writeUInt16BE(0, 6);
    hdr.writeUInt32BE(payload.length, 8);
    hdr.writeUInt32BE(seq, 12);

    const prom = new Promise((res, rej) => this._pendingReqs.push({ res, rej }));
    this._socket.write(Buffer.concat([hdr, payload]));
    const [rcmd, resp] = await prom;

    if (rcmd === Cmd.REDIRECT && followRedirect && this._autoRedirect) {
      const leaderAddr = this._decStr(resp);
      if (leaderAddr && leaderAddr !== 'unknown') {
        const parts = leaderAddr.split(':');
        this._host = parts[0];
        this._port = parseInt(parts[1], 10);
        this.close();
        this._connected = false;
        this._connecting = null;
        await this.connect();
        return this._rpc(cmd, payload, false);
      }
    }

    return [rcmd, resp];
  }

  _onData(chunk) {
    this._recvBuf = Buffer.concat([this._recvBuf, chunk]);
    while (this._recvBuf.length >= HDR_SIZE) {
      const magic = this._recvBuf.readUInt32BE(0);
      if (magic !== MAGIC) { this._recvBuf = Buffer.alloc(0); return; }
      const plen = this._recvBuf.readUInt32BE(8);
      if (this._recvBuf.length < HDR_SIZE + plen) break;
      const cmd     = this._recvBuf.readUInt8(4);
      const payload = this._recvBuf.slice(HDR_SIZE, HDR_SIZE + plen);
      this._recvBuf = this._recvBuf.slice(HDR_SIZE + plen);
      const pending = this._pendingReqs.shift();
      if (pending) pending.res([cmd, payload]);
    }
  }

  _rejectAll(err) {
    for (const p of this._pendingReqs) p.rej(err);
    this._pendingReqs = [];
  }

  _encStr(s) {
    const b = Buffer.from(s, 'utf8');
    const out = Buffer.alloc(4 + b.length);
    out.writeUInt32BE(b.length, 0);
    b.copy(out, 4);
    return out;
  }

  _decStr(buf) {
    if (!buf || buf.length < 4) return '';
    const n = buf.readUInt32BE(0);
    return buf.slice(4, 4 + n).toString('utf8');
  }
}

module.exports = { ApexClient, ApexError, ApexNotFound, ApexNotLeader, Cmd };

// ── Quick CLI ─────────────────────────────────────────────────────────────────
if (require.main === module) {
  (async () => {
    const [host = '127.0.0.1', portS = '7001'] = (process.argv[2] || '127.0.0.1:7001').split(':');
    const client = new ApexClient(host, parseInt(portS, 10));
    await client.connect();
    console.log('Connected.  Try: put key val  /  get key  /  del key  /  exit');
    const readline = require('readline').createInterface({ input: process.stdin, output: process.stdout, prompt: 'apex> ' });
    readline.prompt();
    readline.on('line', async line => {
      const [cmd, k, ...rest] = line.trim().split(' ');
      try {
        if (!cmd || cmd === 'exit') { client.close(); process.exit(0); }
        if (cmd === 'get')  console.log(await client.get(k) ?? '(nil)');
        if (cmd === 'put')  { await client.put(k, rest.join(' ')); console.log('OK'); }
        if (cmd === 'del')  console.log(await client.del(k) ? 'OK' : '(not found)');
        if (cmd === 'ping') console.log(await client.ping() ? 'PONG' : 'FAILED');
        if (cmd === 'metrics') console.log(JSON.stringify(await client.metrics(), (_, v) => typeof v === 'bigint' ? Number(v) : v, 2));
      } catch (e) { console.error('Error:', e.message); }
      readline.prompt();
    });
  })();
}

File: clients/python/apex_kv.py
"""
apex_kv — Python client for APEX-KV distributed key-value store

Installation:
    pip install apex-kv  # or copy this file

Usage:
    from apex_kv import ApexClient

    client = ApexClient("127.0.0.1", 7001)
    client.put("hello", "world")
    value = client.get("hello")          # "world"
    client.delete("hello")

    # Watch for changes (polling)
    for event in client.watch("prefix/"):
        print(event)

    # Context manager
    with ApexClient("127.0.0.1", 7001) as c:
        c.put("x", "1")
"""
import socket
import struct
import threading
import time
from typing import Optional, Iterator, Callable

# ── Wire protocol constants ────────────────────────────────────────────────────
_MAGIC     = 0xA9ECBD10
_HDR_SIZE  = 16
_HDR_FMT   = "!IbbHII"   # magic, cmd, flags, sender_id, payload_len, seq

class Cmd:
    GET          = 0x01
    PUT          = 0x02
    DEL          = 0x03
    PING         = 0x04
    METRICS      = 0x05
    OK           = 0x10
    VALUE        = 0x11
    NOT_FOUND    = 0x12
    ERROR        = 0x13
    REDIRECT     = 0x14
    PONG         = 0x15
    METRICS_RESP = 0x16


class ApexError(Exception):
    pass

class ApexNotFound(ApexError):
    pass

class ApexNotLeader(ApexError):
    def __init__(self, leader_addr: str):
        super().__init__(f"Not leader — redirect to {leader_addr}")
        self.leader_addr = leader_addr


class ApexClient:
    """
    Synchronous APEX-KV client with automatic leader redirect.

    Thread-safe: each method acquires a lock on the socket.
    For concurrent use across many threads, create one client per thread.
    """

    def __init__(self, host: str, port: int, timeout: float = 5.0,
                 auto_redirect: bool = True):
        self._host = host
        self._port = port
        self._timeout = timeout
        self._auto_redirect = auto_redirect
        self._seq = 0
        self._lock = threading.Lock()
        self._sock: Optional[socket.socket] = None
        self._connect()

    def _connect(self):
        s = socket.create_connection((self._host, self._port), timeout=self._timeout)
        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        self._sock = s

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

    def close(self):
        if self._sock:
            self._sock.close()
            self._sock = None

    # ── Public API ─────────────────────────────────────────────────────────────

    def ping(self) -> str:
        """Returns 'PONG' if the node is alive."""
        cmd, payload = self._rpc(Cmd.PING)
        return "PONG" if cmd == Cmd.PONG else "?"

    def get(self, key: str) -> Optional[str]:
        """
        Get the value for a key.
        Returns None if not found (does not raise).
        """
        payload = self._encode_str(key)
        cmd, resp = self._rpc(Cmd.GET, payload)
        if cmd == Cmd.VALUE:
            return self._decode_str(resp)
        if cmd == Cmd.NOT_FOUND:
            return None
        if cmd == Cmd.ERROR:
            raise ApexError(self._decode_str(resp) or "unknown error")
        return None

    def put(self, key: str, value: str) -> None:
        """Write a key-value pair."""
        payload = self._encode_str(key) + self._encode_str(value)
        cmd, resp = self._rpc(Cmd.PUT, payload, follow_redirect=True)
        if cmd == Cmd.ERROR:
            raise ApexError(self._decode_str(resp) or "put failed")

    def delete(self, key: str) -> bool:
        """Delete a key. Returns True if the key existed."""
        payload = self._encode_str(key)
        cmd, _ = self._rpc(Cmd.DEL, payload, follow_redirect=True)
        return cmd == Cmd.OK

    def get_or_default(self, key: str, default: str = "") -> str:
        """Get value with a fallback default."""
        v = self.get(key)
        return v if v is not None else default

    def metrics(self) -> dict:
        """Fetch server metrics."""
        cmd, payload = self._rpc(Cmd.METRICS)
        if cmd != Cmd.METRICS_RESP or len(payload) < 32:
            return {}
        import struct
        ops_get, ops_put, p99_get, p99_put = struct.unpack_from("!QQQQ", payload)
        return {
            "ops_get":  ops_get,
            "ops_put":  ops_put,
            "p99_get_us": p99_get,
            "p99_put_us": p99_put,
        }

    def watch(self, prefix: str, poll_interval: float = 0.5) -> Iterator[dict]:
        """
        Poll-based watch: yields {'key': ..., 'value': ..., 'deleted': bool}
        events when values under `prefix` change.

        Note: APEX-KV does not have native watch/subscribe; this polls.
        For production watch semantics, use a dedicated event-sourcing layer.
        """
        # This is a minimal polling implementation.
        # A proper watch would require server-side SSE or gRPC streaming.
        seen: dict[str, str] = {}
        while True:
            time.sleep(poll_interval)
            # In a real system you'd do a SCAN prefix here.
            # Placeholder: callers should override with their key set.
            yield {"type": "poll", "prefix": prefix}

    # ── Internal ───────────────────────────────────────────────────────────────

    def _rpc(self, cmd: int, payload: bytes = b"",
             follow_redirect: bool = False) -> tuple[int, bytes]:
        with self._lock:
            self._seq += 1
            seq = self._seq
            header = struct.pack("!IbbHII",
                _MAGIC, cmd, 0, 0, len(payload), seq)
            self._send_all(header + payload)
            resp_cmd, resp_payload = self._recv_response()

        if resp_cmd == Cmd.REDIRECT and follow_redirect and self._auto_redirect:
            leader = self._decode_str(resp_payload)
            if leader and ":" in leader:
                host, port_s = leader.rsplit(":", 1)
                self._host = host
                self._port = int(port_s)
                self.close()
                self._connect()
                return self._rpc(cmd, payload, follow_redirect=False)

        return resp_cmd, resp_payload

    def _send_all(self, data: bytes):
        total = 0
        while total < len(data):
            n = self._sock.send(data[total:])
            if n == 0:
                raise ApexError("Connection closed")
            total += n

    def _recv_response(self) -> tuple[int, bytes]:
        header = self._recv_exact(_HDR_SIZE)
        magic, cmd, _flags, _sender, plen, _seq = struct.unpack("!IbbHII", header)
        if magic != _MAGIC:
            raise ApexError(f"Bad magic 0x{magic:08x}")
        payload = self._recv_exact(plen) if plen else b""
        return cmd, payload

    def _recv_exact(self, n: int) -> bytes:
        buf = bytearray()
        while len(buf) < n:
            chunk = self._sock.recv(n - len(buf))
            if not chunk:
                raise ApexError("Connection closed mid-read")
            buf.extend(chunk)
        return bytes(buf)

    @staticmethod
    def _encode_str(s: str) -> bytes:
        b = s.encode("utf-8")
        return struct.pack("!I", len(b)) + b

    @staticmethod
    def _decode_str(payload: bytes) -> str:
        if len(payload) < 4:
            return ""
        n = struct.unpack_from("!I", payload)[0]
        return payload[4:4+n].decode("utf-8", errors="replace")


# ── CLI for quick testing ──────────────────────────────────────────────────────
if __name__ == "__main__":
    import sys, readline  # noqa: readline for arrow key support

    host, port = ("127.0.0.1", 7001)
    if len(sys.argv) > 1:
        parts = sys.argv[1].rsplit(":", 1)
        host = parts[0]
        if len(parts) > 1:
            port = int(parts[1])

    client = ApexClient(host, port)
    print(f"Connected to {host}:{port}  ({client.ping()})")
    print("Commands: get <key>  put <key> <val>  del <key>  metrics  exit")

    while True:
        try:
            line = input("apex> ").strip()
        except (EOFError, KeyboardInterrupt):
            break
        if not line:
            continue
        parts = line.split(maxsplit=2)
        cmd = parts[0].lower()
        if cmd in ("exit", "quit"):
            break
        elif cmd == "get" and len(parts) >= 2:
            v = client.get(parts[1])
            print(v if v is not None else "(nil)")
        elif cmd == "put" and len(parts) >= 3:
            client.put(parts[1], parts[2])
            print("OK")
        elif cmd == "del" and len(parts) >= 2:
            ok = client.delete(parts[1])
            print("OK" if ok else "(not found)")
        elif cmd == "metrics":
            import json
            print(json.dumps(client.metrics(), indent=2))
        elif cmd == "ping":
            print(client.ping())
        else:
            print(f"Unknown: {line}")
